stdlib/public/runtime/Backtrace.cpp

```diff
@@ -789,20 +843,100 @@ namespace backtrace {
 /// @param mangledName is the symbol name to be tested.
 ///
 /// @returns `true` if `mangledName` represents a thunk function.
-SWIFT_RUNTIME_STDLIB_SPI SWIFT_CC(swift) bool
-_swift_isThunkFunction(const char *mangledName) {
+SWIFT_RUNTIME_STDLIB_SPI bool
+_swift_backtrace_isThunkFunction(const char *mangledName) {
   swift::Demangle::Context ctx;
 
   return ctx.isThunkSymbol(mangledName);
 }
 
+// Try to demangle a symbol.
+SWIFT_RUNTIME_STDLIB_SPI char *
+_swift_backtrace_demangle(const char *mangledName,
+                          size_t mangledNameLength,
+                          char *outputBuffer,
+                          size_t *outputBufferSize,
+                          int *status) {
+  llvm::StringRef name = llvm::StringRef(mangledName, mangledNameLength);
+
+  // You must provide buffer size if you're providing your own output buffer
+  if (outputBuffer && !outputBufferSize) {
+    return nullptr;
+  }
+
+  if (Demangle::isSwiftSymbol(name)) {
+    // This is a Swift mangling
+    auto options = DemangleOptions::SimplifiedUIDemangleOptions();
+    auto result = Demangle::demangleSymbolAsString(name, options);
+    size_t bufferSize;
+
+    if (outputBufferSize) {
+      bufferSize = *outputBufferSize;
+      *outputBufferSize = result.length() + 1;
+    }
+
+    if (outputBuffer == nullptr) {
+      outputBuffer = (char *)::malloc(result.length() + 1);
+      bufferSize = result.length() + 1;
+    }
+
+    size_t toCopy = std::min(bufferSize - 1, result.length());
+    ::memcpy(outputBuffer, result.data(), toCopy);
+    outputBuffer[toCopy] = '\0';
+
+    *status = 0;
+    return outputBuffer;
+#ifndef _WIN32
+  } else if (name.startswith("_Z")) {
+    // Try C++
+    size_t resultLen;
+    char *result = abi::__cxa_demangle(mangledName, nullptr, &resultLen, status);
```

Comment:
Doesn't `abi::__cxa_demangle` pretty much define the same behavior? Do we need the wrapper checks remapping the results? (It doesn't check the mangledNameLength, which I would appreciate)
But something like this
`return abi::__cxa_demangle(mangledName, outputBuffer, outputBufferSize, status);` looks like it should do the same thing, but without the additional memcpy. I suppose this doesn't work if the `outputBuffer` isn't allocated with `malloc` though. :thinking:. (If that's the reason why we're stashing the pointer and copying the result manually, that would be good to comment so someone doesn't come along and try reducing it.)

Which reminds me, we should probably document that the result pointer needs to be free'd with `free` in either case.

Reply:
Yeah, I did that so that you didn't have to use `malloc()` to allocate the buffer. One of the nasty things about the `abi::__cxa_demangle` function is that it requires that the pointer you pass in is `malloc()`ed and might even `realloc()` it for you. I can see how that came about, but I can't say I'm keen on it as an interface.

Reply:
yeah, okay. Using `realloc` internally.... That's an interesting decision. Not a huge fan, but oh well.

Reply:
Anyway comment if you could add a comment to the call location, I'm fine with this how it is then.